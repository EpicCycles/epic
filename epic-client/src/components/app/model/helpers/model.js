import {
    findObjectWithKey,
    generateRandomCode,
    isItAnObject,
    removeKey,
    updateObject,
    updateObjectInArray
} from "../../../../helpers/utils";
import {NEW_ELEMENT_ID} from "../../../../helpers/constants";

export const eliminateReadOnlyFields = (fieldList) => {
    return fieldList.filter(field => (! field.readOnly));
};

export const checkForChanges = (fieldList, existingObject, newValues) => {
    return eliminateReadOnlyFields(fieldList).some(field => {
        return existingObject[field.fieldName] !== newValues[field.fieldName];
    })
};

export const addFieldToState = (initialState, fieldList, fieldName, input) => {
    let finalState = initialState;
    fieldList.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            finalState = applyFieldValueToModel(initialState, field, input);
            return true;
        }
        return false;
    });
    return finalState;
};


export const getModelKey = modelInstance => {
    if (!modelInstance) return NEW_ELEMENT_ID;
    if (modelInstance.id) return modelInstance.id;
    if (modelInstance.dummyKey) return modelInstance.dummyKey;
    return NEW_ELEMENT_ID;
};

export const isModelValid = (modelInstance) => {
    return !(modelInstance.error || isItAnObject(modelInstance.error_detail));
};

export const applyFieldValueToModel = (modelInstance, field, value) => {
    let updatedModelInstance = updateObject(modelInstance);
    updatedModelInstance[field.fieldName] = value;
    updatedModelInstance.changed = true;
    if (!updatedModelInstance.error_detail) updatedModelInstance.error_detail = {};
    if (field.required && !value) {
        updatedModelInstance.error_detail[field.fieldName] = field.error;
        return updatedModelInstance;
    } else if (field.validator) {
        const error = field.validator(value, modelInstance);
        if (error) {
            updatedModelInstance.error_detail[field.fieldName] = error;
            return updatedModelInstance;
        }
    }
    updatedModelInstance.error_detail = removeKey(updatedModelInstance.error_detail, field.fieldName);
    return updatedModelInstance;
};

export const getAttribute = (modelFields, fieldName) => {
    let attribute;
    modelFields.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            attribute = field.fieldName;
            return true;
        }
        return false;
    });
    return attribute;
};

export const getField = (modelFields, fieldName) => {
    let modelField;
    modelFields.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            modelField = field;
            return true;
        }
        return false;
    });
    return modelField;
};

export const updateModelChanges = (modelInstance, attribute, fieldValue) => {
    let changes = updateObject(modelInstance.changes);
    if (fieldValue && (modelInstance[attribute] !== fieldValue)) {
        changes[attribute] = fieldValue;
    } else {
        changes = removeKey(changes, attribute);
    }
    return updateObject(modelInstance, { changes })
};
export const updateModelWithChanges = (modelInstance, modelFields, fieldName, fieldValue) => {
    const attribute = getAttribute(modelFields, fieldName);
    if (attribute) {
        return updateModelChanges(modelInstance, attribute, fieldValue)
    } else {
        console.error("attribute not found for", modelFields, fieldName)
    }
    return modelInstance;
};
export const updateModelArrayWithChanges = (modelArray, modelFields, fieldName, fieldValue, componentKey) => {
    const attribute = getAttribute(modelFields, fieldName);
    if (attribute) {
        let modelInstance = findObjectWithKey(modelArray, componentKey);
        let currentChanges = modelInstance.changes || {};
        if (fieldValue && (modelInstance[attribute] !== fieldValue)) {
            currentChanges[attribute] = fieldValue;
        } else {
            currentChanges = removeKey(currentChanges, attribute);
        }
        modelInstance.changes = currentChanges;
        return updateObjectInArray(modelArray, modelInstance, componentKey)
    } else {
        console.error("attribute not found for", modelFields, fieldName)
    }
    return modelArray;
};
export const updateModel = (model, modelFields, fieldName, fieldValue, componentKey) => {
    const modelField = getField(modelFields, fieldName);
    if (modelField) return applyFieldValueToModel(model, modelField, fieldValue);
    return model;
};

export const displayModelErrorSummary = (model, modelFields) => {
    let displayErrors = model.error;
    const fieldErrors = isItAnObject(model.error_detail) ? model.error_detail : {};
    for (var property in fieldErrors) {
        const field = getField(modelFields, property);
        if (!displayErrors) {
            displayErrors = "";
        } else {
            displayErrors += '<br>';
        }
        displayErrors += field.header + ": " + fieldErrors[property].join(" ");
    }

    return displayErrors;
};

export const getNameForValue = (value, sourceArray) => {
    if (!value) return undefined;
    let name = value;

    sourceArray.some(pair => {
        if (pair.value === value) {
            name = pair.name;
            return true;
        }
        return false;
    });
    return name;
};

export const hasErrors = (model) => {
    if (model.error && model.error.length > 0) return true;
    if (isItAnObject(model.error_detail)) return true;
    return false;
};

export const addErrorDetail = (error_detail = {}, field, error) => {
      let updated_error_detail = updateObject(error_detail);
  if (!Array.isArray(updated_error_detail[field])) {
         updated_error_detail[field] = [];
  }
  updated_error_detail[field].push(error);
  return updated_error_detail;
};

export const createNewModelInstance = () => {
    return {dummyKey: generateRandomCode()};
};
export const matchesModel = (persistedModel, modelFields, modelToCheck) => {
     return eliminateReadOnlyFields(modelFields).every(field => {
        return persistedModel[field.fieldName] === modelToCheck[field.fieldName];
    })
};