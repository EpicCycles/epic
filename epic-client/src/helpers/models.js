export const BRAND = "brand";
export const FRAME_NAME = "frame_name";
export const MODEL_NAME = "model_name";
export const DESCRIPTION = "description";
export const COLOURS = "colours";
export const SELL_PRICE = "sell_price";
export const SIZES = "sizes";

export const BRAND_FIELD = { fieldName: BRAND, header: "Brand", synonyms: [BRAND, 'manufacturer'] };
export const FRAME_NAME_FIELD = { fieldName: FRAME_NAME, header: "Frame Name" };
export const MODEL_NAME_FIELD = { fieldName: MODEL_NAME, header: "Model Name" };
export const COLOURS_FIELD = {
    fieldName: COLOURS,
    header: "Colours",
    synonyms: [COLOURS, "colour", "colors", "color"]
};
export const DESCRIPTION_FIELD = { fieldName: DESCRIPTION, header: "Description", synonyms: [DESCRIPTION, "desc"] };
export const SELL_PRICE_FIELD = {
    fieldName: SELL_PRICE,
    header: "Selling Price",
    synonyms: ["selling price", "srp", "rrp", "sell price", "retail price"]
};
export const SIZES_FIELD = {
    fieldName: SIZES,
    header: "Sizes",
    synonyms: [SIZES, "size", "frame sizes", "frame size"]
};
export const frameFields = [BRAND, FRAME_NAME];
export const bikeFields = [MODEL_NAME_FIELD, DESCRIPTION_FIELD, COLOURS_FIELD, SELL_PRICE_FIELD, SIZES_FIELD];