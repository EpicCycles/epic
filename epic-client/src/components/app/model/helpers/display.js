export const fixedHeaderClassname = (lockColumn) => {
    if (lockColumn) return "grid-header--fixed-left";
    return "";
};
export const fixedDetailsClassname = (lockColumn) => {
    if (lockColumn) return "grid-item--fixed-left";
    return "";
};
export const gridHeaderClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item--header ${fixedHeaderClassname(shouldLock)}`;
};
export const gridItemClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item ${fixedDetailsClassname(shouldLock)}`;
};