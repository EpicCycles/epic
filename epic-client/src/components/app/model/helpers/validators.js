import {INVALID_POSTCODE} from "./error";
import {POSTCODE_RULES} from "./constants";
export const validateData = (fieldList, currentValues) => {
    let errors = {};
    fieldList.forEach(field => {
        if (field.required) {
            if (!currentValues[field.fieldName]) {
                errors[field.fieldName] = field.error;
            }
        }
    });
    return errors;
};

export const validateEmailFormat = (email) => {
    const emailPattern = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return emailPattern.test(email);
};

export const validatePostcodeAndReturnError = (postcode, model = {}) => {
    let error;
    const postcodeRule = POSTCODE_RULES.filter(rule => rule.countryCode === model.country);
    if (postcodeRule[0]) {
        const OK = RegExp(postcodeRule[0].regex).exec(postcode);
        if (!OK) error = INVALID_POSTCODE + postcodeRule[0].display;
    }
    return error;
};