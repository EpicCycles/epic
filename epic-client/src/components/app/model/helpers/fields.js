import {
    ADDRESS_MISSING,
    BRAND_MISSING, BRAND_NAME_MISSING,
    BUNDLE_NAME_MISSING, COUNTRY_MISSING,
    FRAME_NAME_MISSING,
    MODEL_NAME_MISSING,
    PART_MISSING,
    PART_NAME_MISSING,
    PART_TYPE_MISSING,
    PRODUCTS_MISSING,
    SUPPLIER_MISSING
} from "./error";
import {validatePostcodeAndReturnError} from "./validators";

// TODO customer fields including email with email validator
export const ADDRESS1 = "address1";
export const ADDRESS2 = "address2";
export const ADDRESS3 = "address3";
export const ADDRESS4 = "address4";
export const COUNTRY = "country";
export const POSTCODE = "postcode";
export const BRAND = "brand";
export const SUPPLIER = "supplier";
export const PART = "part";
export const LINK = "link";
export const BUNDLE = "bundle_name";
export const PRODUCT_CODE = "product_code";
export const PART_TYPE = "partType";
export const BRAND_NAME = "brand_name";
export const BIKE_BRAND = "bike_brand";
export const FRAME_NAME = "frame_name";
export const MODEL_NAME = "model_name";
export const PART_NAME = "part_name";
export const DESCRIPTION = "description";
export const COLOURS = "colours";
export const SELL_PRICE = "rrp";
export const EPIC_PRICE = "epic_price";
export const TICKET_PRICE = "ticket_price";
export const TRADE_PRICE = "trade_price";
export const CLUB_PRICE = "club_price";
export const FITTED_PRICE = "fitted_price";
export const TRADE_IN_PRICE = "trade_in_value";
export const SIZES = "sizes";
export const STANDARD = "standard";
export const STOCKED = "stocked";
export const PRODUCTS = "products";
export const TEXT = "text";
export const NUMBER = "number";
export const CURRENCY = "currency";
export const TEXT_AREA = "textArea";
export const CHECKBOX = "checkbox";

// TODO add COUNTRY type and fields for address
// TODO consider having a function passed for validator
export const ADDRESS1_FIELD = {
    fieldName: ADDRESS1,
    type: TEXT,
    length: 100,
    header: "Address",
    required: true,
    error: ADDRESS_MISSING
};
export const ADDRESS2_FIELD = {
    fieldName: ADDRESS2,
    type: TEXT,
    length: 100,
    header: "Line 2",
};
export const ADDRESS3_FIELD = {
    fieldName: ADDRESS3,
    type: TEXT,
    length: 100,
    header: "Line 3",
};
export const ADDRESS4_FIELD = {
    fieldName: ADDRESS4,
    type: TEXT,
    length: 100,
    header: "Line 4",
};
export const COUNTRY_FIELD = {
    fieldName: COUNTRY,
    header: "Country",
    type: COUNTRY,
    required: true,
    error: COUNTRY_MISSING
};
export const POSTCODE_FIELD = {
    fieldName: POSTCODE,
    type: TEXT,
    length: 20,
    header: "Postcode",
    validator: validatePostcodeAndReturnError,
    validatorAdditionalFields: [COUNTRY]
};
export const BRAND_FIELD = {
    fieldName: BRAND,
    header: "Brand",
    synonyms: [BRAND, 'manufacturer'],
    required: true,
    error: BRAND_MISSING,
    type: BRAND
};
export const PRODUCTS_FIELD = {
    fieldName: PRODUCTS,
    header: "Products",
    synonyms: [],
    required: true,
    error: PRODUCTS_MISSING,
    type: PRODUCTS
};
export const SUPPLIER_FIELD = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    required: true,
    error: SUPPLIER_MISSING,
    type: SUPPLIER
};
export const SUPPLIER_FIELD_OPTIONAL = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    type: SUPPLIER
};
export const PART_FIELD = {
    fieldName: PART,
    header: "Part",
    synonyms: [PART],
    required: true,
    error: PART_MISSING,
    type: PART
};
export const FRAME_NAME_FIELD = {
    fieldName: FRAME_NAME,
    header: "Frame Name",
    synonyms: [],
    required: true,
    error: FRAME_NAME_MISSING
};
export const LINK_FIELD = {
    fieldName: LINK,
    header: "URL",
    synonyms: [],
};
export const BRAND_NAME_FIELD = {
    fieldName: BRAND_NAME,
    header: "Brand Name",
    synonyms: [],
    required: true,
    error: BRAND_NAME_MISSING
};
export const BUNDLE_NAME_FIELD = {
    fieldName: BUNDLE,
    header: "Bundle Name",
    synonyms: [],
    required: true,
    error: BUNDLE_NAME_MISSING
};
export const MODEL_NAME_FIELD = {
    fieldName: MODEL_NAME,
    header: "Model Name",
    synonyms: [],
    required: true,
    error: MODEL_NAME_MISSING,
    type: TEXT,
    length: 100
};
export const COLOURS_FIELD = {
    fieldName: COLOURS,
    header: "Colours",
    synonyms: [COLOURS, "colour", "colors", "color"],
    type: TEXT,
    length: 100
};
export const DESCRIPTION_FIELD = {
    fieldName: DESCRIPTION,
    header: "Description",
    synonyms: [DESCRIPTION, "desc"],
    type: TEXT_AREA,
    length: 400
};
export const PRODUCT_CODE_FIELD = {
    fieldName: PRODUCT_CODE,
    header: "Product Code",
    synonyms: [PRODUCT_CODE],
    type: TEXT,
    length: 30
};
export const SELL_PRICE_FIELD = {
    fieldName: SELL_PRICE,
    header: "RRP",
    synonyms: ["price", "selling price", "srp", "rrp", "sell price", "retail price"],
    type: CURRENCY,
    length: 10
};
export const EPIC_PRICE_FIELD = {
    fieldName: EPIC_PRICE,
    header: "Epic Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const CLUB_PRICE_FIELD = {
    fieldName: CLUB_PRICE,
    header: "Club Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};

export const FITTED_PRICE_FIELD = {
    fieldName: FITTED_PRICE,
    header: "Fitted Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TRADE_PRICE_FIELD = {
    fieldName: TRADE_PRICE,
    header: "Trade Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TICKET_PRICE_FIELD = {
    fieldName: TICKET_PRICE,
    header: "Ticket Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const SIZES_FIELD = {
    fieldName: SIZES,
    header: "Sizes",
    synonyms: [SIZES, "size", "frame sizes", "frame size"],
    type: TEXT,
    length: 100
};
export const PART_TYPE_FIELD = {
    fieldName: PART_TYPE,
    header: "Part Type",
    required: true,
    error: PART_TYPE_MISSING,
    type: PART_TYPE,
};
export const PART_NAME_FIELD = {
    fieldName: PART_NAME,
    header: "Part Name",
    required: true,
    error: PART_NAME_MISSING
};
export const TRADE_IN_FIELD = {
    fieldName: TRADE_IN_PRICE,
    header: "Trade In Â£",
    type: CURRENCY,
    length: 10
};
export const STOCKED_FIELD = {
    fieldName: STOCKED,
    header: "Stocked",
    type: CHECKBOX
};
export const STANDARD_FIELD = {
    fieldName: STANDARD,
    header: "Standard",
    type: CHECKBOX
};
export const BIKE_BRAND_FIELD = {
    fieldName: BIKE_BRAND,
    header: "Bike Brand",
    type: CHECKBOX
};
export const customerAddressFields = [ADDRESS1_FIELD, ADDRESS2_FIELD, ADDRESS3_FIELD, ADDRESS4_FIELD, COUNTRY_FIELD, POSTCODE_FIELD];
export const frameFields = [BRAND, FRAME_NAME];
export const brandFields = [BRAND_NAME_FIELD, BIKE_BRAND_FIELD, LINK_FIELD, SUPPLIER_FIELD_OPTIONAL];
export const bikeFields = [MODEL_NAME_FIELD, DESCRIPTION_FIELD, COLOURS_FIELD, SELL_PRICE_FIELD, EPIC_PRICE_FIELD, CLUB_PRICE_FIELD, SIZES_FIELD];
export const partFields = [
    PART_TYPE_FIELD,
    BRAND_FIELD,
    PART_NAME_FIELD,
    TRADE_IN_FIELD,
    STANDARD_FIELD,
    STOCKED_FIELD
];
export const supplierProductFields = [
    SUPPLIER_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
    SELL_PRICE_FIELD,
    TRADE_PRICE_FIELD,
    CLUB_PRICE_FIELD
];
export const bundleFields = [
    BUNDLE_NAME_FIELD,
    PRODUCTS_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
];

