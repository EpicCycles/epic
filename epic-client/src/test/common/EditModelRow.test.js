import React from "react";
import EditModelRow from "../../common/EditModelRow";
import {customerAddressFields} from "../../helpers/models";

const foundName = "find me";
const sections = [
    {
        id: 1,
        partTypes: [
            { id: 11, shortName: "id 11" },
            { id: 21, shortName: "id 11" },
        ]
    },
    {
        id: 2,
        partTypes: [
            { id: 2, shortName: foundName },
            { id: 22, shortName: "id 11" },
        ]
    },
];
const brands = [
    { id: 1, brand_name: "id is 1" },
    { id: 2, brand_name: foundName },
    { id: 3, brand_name: "id is 3" },
];
const suppliers = [
    { id: 1, supplier_name: "id is 1" },
    { id: 2, supplier_name: foundName },
    { id: 3, supplier_name: "id is 3" },
];
const emptyModel = {};
const model = {
    id: 123,
    address1: "line one",
    address2: "line Ywo",
    address3: "line Three",
    address4: "line Four",
    postcode: "xxxyyy",
    customer: 6
};
test("it renders", () => {
    const component = shallow(<EditModelRow model={emptyModel} modelFields={customerAddressFields}
                                            onChange={jest.fn()}/>);
    expect(component).toMatchSnapshot();
})
test("it renders with data", () => {
    const component = shallow(<EditModelRow
        model={model}
        modelFields={customerAddressFields}
        onChange={jest.fn()}
        className="red"
        suppliers={suppliers}
        sections={sections}
        brands={brands}
        persistedModel={model}
        lockFirstColumn={true}
        childModels={[{ id: 23 }, { id: 34 }]}
    />);
    expect(component).toMatchSnapshot();
})