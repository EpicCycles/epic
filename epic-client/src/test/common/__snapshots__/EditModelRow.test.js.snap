// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<React.Fragment>
  <div
    className="grid-item"
    key="modelRowaddress1new"
    style={
      Object {
        "gridRow": "span 1;",
      }
    }
  >
    <EditModelInput
      componentKey="new"
      field={
        Object {
          "error": "At least one line of address must be entered",
          "fieldName": "address1",
          "header": "Address",
          "length": 100,
          "required": true,
          "type": "text",
        }
      }
      index={0}
      model={Object {}}
      onChange={[MockFunction]}
    />
  </div>
  <div
    className="grid-item"
    key="modelRowaddress2new"
    style={
      Object {
        "gridRow": "span 1;",
      }
    }
  >
    <EditModelInput
      componentKey="new"
      field={
        Object {
          "fieldName": "address2",
          "header": "Line 2",
          "length": 100,
          "type": "text",
        }
      }
      index={1}
      model={Object {}}
      onChange={[MockFunction]}
    />
  </div>
  <div
    className="grid-item"
    key="modelRowaddress3new"
    style={
      Object {
        "gridRow": "span 1;",
      }
    }
  >
    <EditModelInput
      componentKey="new"
      field={
        Object {
          "fieldName": "address3",
          "header": "Line 3",
          "length": 100,
          "type": "text",
        }
      }
      index={2}
      model={Object {}}
      onChange={[MockFunction]}
    />
  </div>
  <div
    className="grid-item"
    key="modelRowaddress4new"
    style={
      Object {
        "gridRow": "span 1;",
      }
    }
  >
    <EditModelInput
      componentKey="new"
      field={
        Object {
          "fieldName": "address4",
          "header": "Line 4",
          "length": 100,
          "type": "text",
        }
      }
      index={3}
      model={Object {}}
      onChange={[MockFunction]}
    />
  </div>
  <div
    className="grid-item"
    key="modelRowcountrynew"
    style={
      Object {
        "gridRow": "span 1;",
      }
    }
  >
    <EditModelInput
      componentKey="new"
      field={
        Object {
          "error": "A country must be selected",
          "fieldName": "country",
          "header": "Country",
          "required": true,
          "type": "country",
        }
      }
      index={4}
      model={Object {}}
      onChange={[MockFunction]}
    />
  </div>
  <div
    className="grid-item"
    key="modelRowpostcodenew"
    style={
      Object {
        "gridRow": "span 1;",
      }
    }
  >
    <EditModelInput
      componentKey="new"
      field={
        Object {
          "fieldName": "postcode",
          "header": "Postcode",
          "length": 20,
          "type": "text",
          "validator": [Function],
          "validatorAdditionalFields": Array [
            "country",
          ],
        }
      }
      index={5}
      model={Object {}}
      onChange={[MockFunction]}
    />
  </div>
</React.Fragment>
`;

exports[`it renders with data 1`] = `
<React.Fragment>
  <div
    className="grid-item grid-item--fixed-leftred"
    key="modelRowaddress1123"
    style={
      Object {
        "gridRow": "span 2;",
      }
    }
  >
    <EditModelInput
      brands={
        Array [
          Object {
            "brand_name": "id is 1",
            "id": 1,
          },
          Object {
            "brand_name": "find me",
            "id": 2,
          },
          Object {
            "brand_name": "id is 3",
            "id": 3,
          },
        ]
      }
      componentKey={123}
      field={
        Object {
          "error": "At least one line of address must be entered",
          "fieldName": "address1",
          "header": "Address",
          "length": 100,
          "required": true,
          "type": "text",
        }
      }
      index={0}
      model={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      onChange={[MockFunction]}
      persistedModel={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      sections={
        Array [
          Object {
            "id": 1,
            "partTypes": Array [
              Object {
                "id": 11,
                "shortName": "id 11",
              },
              Object {
                "id": 21,
                "shortName": "id 11",
              },
            ],
          },
          Object {
            "id": 2,
            "partTypes": Array [
              Object {
                "id": 2,
                "shortName": "find me",
              },
              Object {
                "id": 22,
                "shortName": "id 11",
              },
            ],
          },
        ]
      }
      suppliers={
        Array [
          Object {
            "id": 1,
            "supplier_name": "id is 1",
          },
          Object {
            "id": 2,
            "supplier_name": "find me",
          },
          Object {
            "id": 3,
            "supplier_name": "id is 3",
          },
        ]
      }
    />
  </div>
  <div
    className="grid-item grid-item--fixed-leftred"
    key="modelRowaddress2123"
    style={
      Object {
        "gridRow": "span 2;",
      }
    }
  >
    <EditModelInput
      brands={
        Array [
          Object {
            "brand_name": "id is 1",
            "id": 1,
          },
          Object {
            "brand_name": "find me",
            "id": 2,
          },
          Object {
            "brand_name": "id is 3",
            "id": 3,
          },
        ]
      }
      componentKey={123}
      field={
        Object {
          "fieldName": "address2",
          "header": "Line 2",
          "length": 100,
          "type": "text",
        }
      }
      index={1}
      model={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      onChange={[MockFunction]}
      persistedModel={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      sections={
        Array [
          Object {
            "id": 1,
            "partTypes": Array [
              Object {
                "id": 11,
                "shortName": "id 11",
              },
              Object {
                "id": 21,
                "shortName": "id 11",
              },
            ],
          },
          Object {
            "id": 2,
            "partTypes": Array [
              Object {
                "id": 2,
                "shortName": "find me",
              },
              Object {
                "id": 22,
                "shortName": "id 11",
              },
            ],
          },
        ]
      }
      suppliers={
        Array [
          Object {
            "id": 1,
            "supplier_name": "id is 1",
          },
          Object {
            "id": 2,
            "supplier_name": "find me",
          },
          Object {
            "id": 3,
            "supplier_name": "id is 3",
          },
        ]
      }
    />
  </div>
  <div
    className="grid-item grid-item--fixed-leftred"
    key="modelRowaddress3123"
    style={
      Object {
        "gridRow": "span 2;",
      }
    }
  >
    <EditModelInput
      brands={
        Array [
          Object {
            "brand_name": "id is 1",
            "id": 1,
          },
          Object {
            "brand_name": "find me",
            "id": 2,
          },
          Object {
            "brand_name": "id is 3",
            "id": 3,
          },
        ]
      }
      componentKey={123}
      field={
        Object {
          "fieldName": "address3",
          "header": "Line 3",
          "length": 100,
          "type": "text",
        }
      }
      index={2}
      model={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      onChange={[MockFunction]}
      persistedModel={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      sections={
        Array [
          Object {
            "id": 1,
            "partTypes": Array [
              Object {
                "id": 11,
                "shortName": "id 11",
              },
              Object {
                "id": 21,
                "shortName": "id 11",
              },
            ],
          },
          Object {
            "id": 2,
            "partTypes": Array [
              Object {
                "id": 2,
                "shortName": "find me",
              },
              Object {
                "id": 22,
                "shortName": "id 11",
              },
            ],
          },
        ]
      }
      suppliers={
        Array [
          Object {
            "id": 1,
            "supplier_name": "id is 1",
          },
          Object {
            "id": 2,
            "supplier_name": "find me",
          },
          Object {
            "id": 3,
            "supplier_name": "id is 3",
          },
        ]
      }
    />
  </div>
  <div
    className="grid-item grid-item--fixed-leftred"
    key="modelRowaddress4123"
    style={
      Object {
        "gridRow": "span 2;",
      }
    }
  >
    <EditModelInput
      brands={
        Array [
          Object {
            "brand_name": "id is 1",
            "id": 1,
          },
          Object {
            "brand_name": "find me",
            "id": 2,
          },
          Object {
            "brand_name": "id is 3",
            "id": 3,
          },
        ]
      }
      componentKey={123}
      field={
        Object {
          "fieldName": "address4",
          "header": "Line 4",
          "length": 100,
          "type": "text",
        }
      }
      index={3}
      model={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      onChange={[MockFunction]}
      persistedModel={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      sections={
        Array [
          Object {
            "id": 1,
            "partTypes": Array [
              Object {
                "id": 11,
                "shortName": "id 11",
              },
              Object {
                "id": 21,
                "shortName": "id 11",
              },
            ],
          },
          Object {
            "id": 2,
            "partTypes": Array [
              Object {
                "id": 2,
                "shortName": "find me",
              },
              Object {
                "id": 22,
                "shortName": "id 11",
              },
            ],
          },
        ]
      }
      suppliers={
        Array [
          Object {
            "id": 1,
            "supplier_name": "id is 1",
          },
          Object {
            "id": 2,
            "supplier_name": "find me",
          },
          Object {
            "id": 3,
            "supplier_name": "id is 3",
          },
        ]
      }
    />
  </div>
  <div
    className="grid-item grid-item--fixed-leftred"
    key="modelRowcountry123"
    style={
      Object {
        "gridRow": "span 2;",
      }
    }
  >
    <EditModelInput
      brands={
        Array [
          Object {
            "brand_name": "id is 1",
            "id": 1,
          },
          Object {
            "brand_name": "find me",
            "id": 2,
          },
          Object {
            "brand_name": "id is 3",
            "id": 3,
          },
        ]
      }
      componentKey={123}
      field={
        Object {
          "error": "A country must be selected",
          "fieldName": "country",
          "header": "Country",
          "required": true,
          "type": "country",
        }
      }
      index={4}
      model={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      onChange={[MockFunction]}
      persistedModel={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      sections={
        Array [
          Object {
            "id": 1,
            "partTypes": Array [
              Object {
                "id": 11,
                "shortName": "id 11",
              },
              Object {
                "id": 21,
                "shortName": "id 11",
              },
            ],
          },
          Object {
            "id": 2,
            "partTypes": Array [
              Object {
                "id": 2,
                "shortName": "find me",
              },
              Object {
                "id": 22,
                "shortName": "id 11",
              },
            ],
          },
        ]
      }
      suppliers={
        Array [
          Object {
            "id": 1,
            "supplier_name": "id is 1",
          },
          Object {
            "id": 2,
            "supplier_name": "find me",
          },
          Object {
            "id": 3,
            "supplier_name": "id is 3",
          },
        ]
      }
    />
  </div>
  <div
    className="grid-item grid-item--fixed-leftred"
    key="modelRowpostcode123"
    style={
      Object {
        "gridRow": "span 2;",
      }
    }
  >
    <EditModelInput
      brands={
        Array [
          Object {
            "brand_name": "id is 1",
            "id": 1,
          },
          Object {
            "brand_name": "find me",
            "id": 2,
          },
          Object {
            "brand_name": "id is 3",
            "id": 3,
          },
        ]
      }
      componentKey={123}
      field={
        Object {
          "fieldName": "postcode",
          "header": "Postcode",
          "length": 20,
          "type": "text",
          "validator": [Function],
          "validatorAdditionalFields": Array [
            "country",
          ],
        }
      }
      index={5}
      model={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      onChange={[MockFunction]}
      persistedModel={
        Object {
          "address1": "line one",
          "address2": "line Ywo",
          "address3": "line Three",
          "address4": "line Four",
          "customer": 6,
          "id": 123,
          "postcode": "xxxyyy",
        }
      }
      sections={
        Array [
          Object {
            "id": 1,
            "partTypes": Array [
              Object {
                "id": 11,
                "shortName": "id 11",
              },
              Object {
                "id": 21,
                "shortName": "id 11",
              },
            ],
          },
          Object {
            "id": 2,
            "partTypes": Array [
              Object {
                "id": 2,
                "shortName": "find me",
              },
              Object {
                "id": 22,
                "shortName": "id 11",
              },
            ],
          },
        ]
      }
      suppliers={
        Array [
          Object {
            "id": 1,
            "supplier_name": "id is 1",
          },
          Object {
            "id": 2,
            "supplier_name": "find me",
          },
          Object {
            "id": 3,
            "supplier_name": "id is 3",
          },
        ]
      }
    />
  </div>
</React.Fragment>
`;
